Support negative indices in maps

chrysn opened the pull request:
Closes: https://github.com/twittner/minicbor/issues/8

---

An existing test was extended to cover the feature, I think this can just work like that.

This does introduce a theoretical regression in that arrays with more than 32k elements now cause trouble (a very theoretical issue) and that maps with keys between 32k and 64k now fail. If you consider those important, I'm happy to extend the numeric range of map keys to i64.

On the side of error handling, so far for where I've tried things, the first error that cropped up when attempting to use a negative integer in an array was always this, which I think is fine:

```
error: proc-macro derive panicked                                          
  --> minicbor-tests/tests/encoding.rs:62:21      
   |
62 |     #[derive(Debug, Encode, Decode, PartialEq, Eq)]                                                                                              
   |                     ^^^^^^
   |                    
   = help: message: Negative index value can only be used as map key, not as array index.: TryFromIntError(())
```

chrysn commented:
This was easier than expected, thanks to `cargo expand`; I've updated the PR description and think this is ready for review.

twittner commented on minicbor-derive/src/attrs/idx.rs:
```diff
@@ -24,12 +24,24 @@ impl Idx {
+    pub fn val_as_index(self) -> u32 {
+        self.val()
+            .try_into()
+            .expect("negative index value can only be used as map key, not as array index")
```
When collecting the attributes a proper `syn::Error` with a corresponding span should be returned when a negative index is used together with `#[cbor(map)]`. Panicking here is maybe a bit coarse.

chrysn replied:
I only have very little experience with macros: How would I do it better? All call sites can return a syn::Result, but constructing one requires span information. Should that be passed in by the caller, or should the Idx store the span of where it was constructed so that it can return that back out when it fails?

chrysn replied:
I've tried it out with an arbitrary struct that uses default list encoding, putting in a negative number. The error is:

```
error: proc-macro derive panicked
   --> /home/chrysn/git/crates/ariel-os/src/lib/coapcore/src/ace.rs:146:18
    |
146 |         #[derive(minicbor::Encode)]
    |                  ^^^^^^^^^^^^^^^^
    |
    = help: message: negative index value can only be used as map key, not as array index: TryFromIntError(())
```

That's indeed not super pretty, but given that the numbers have to be hardcoded in the definition (no constants supported), it may be good enough.

To make it better, I think I'd need a hint as to how it *should* be done.

twittner replied:
I am a bit busy at the moment so I can not assist right now. Generally, attributes are collected and checked in `minicbor-derive/src/attrs.rs`.

chrysn replied:
That was a good hint, I found a pattern from there. Now the error message is:

```
error: negative index value can only be used as map key, not as array index
  --> minicbor-tests/tests/encoding.rs:67:9
   |
67 |         #[n(-1)] c: Option<u8>
   |         ^
```

As by-catch I've also fixed the error message for overly large numbers -- not a very realistic case, as it'd only happen on 16-bit hosts, but it was an example I should not have followed.

twittner commented on minicbor-derive/src/fields.rs:
```diff
@@ -63,7 +63,7 @@ impl Fields {
         }
 
-        fields.sort_unstable_by_key(|f| f.index.val());
+        fields.sort_unstable_by_key(|f| f.index.val_for_sorting());
```
Why is this required?

chrysn replied:
The comment on the function was incomplete; now fixed.

My understanding of what this sorting does here is to ensure that when serialization happens field by field, the resulting map is in preferred encoding; `val_for_sorting()` creates an order that is identical with the order of the later encoded items (which was trivial for u32, but is not when negative items are allowed).

twittner commented:
My apologies for the delay. I intend to review your PR by the end of this week.

twittner reviewed (commented):
Overall looks good too me. I left some smaller remarks. Did you find `field.orig` gives better error reporting compared to `field.index`?

twittner reviewed (changes requested):
I forgot to mention that your commits are not all signed-off.

chrysn commented on minicbor-derive/src/attrs/idx.rs:
```diff
@@ -24,12 +24,30 @@ impl Idx {
+    /// Get a property of the index that is sorted as needed to encoded CBOR map keys
+    /// deterministically (that is, sorted like their deterministic encoding).
+    ///
+    /// This is used for pre-sorting map items for serialization.
```
Better like this? (from d3547534c82f552d8b12d3c898ea32301be8710c):

```rust
    /// Get a property of the index that is sorted as needed to produce the encoded items:
    /// They are sorted in their deterministic encoding.
    ///
    /// This produces the correct sequence both for CBOR arrays (where all items are emitted in
    /// ascending order) and for CBOR maps (where they keys need to be sorted in bytewise
    /// lexicographic ordering).
    ///
    /// This is used for pre-sorting items for serialization.
```

twittner replied:
"where ~~they~~ keys"

chrysn replied:
Thanks, fixed; starting a rebase. That version will have the same tree with ordered and signed-off commits.

chrysn commented:
> Did you find field.orig gives better error reporting compared to field.index?

Yes: `.index` is just an `Idx`, and that merely carries the value. Its implementation of `Spanend` (which provides `.span()`) comes from `ToTokens`, and that creates a token tree out of thin air (what else should it do: the data is not available on that item).

In particular, with the suggestions applied as originally, I get this first error when I remove the `#[cbor(map)]` from line 63 of minicbor-tests/tests/encoding.rs:

```
error: array encoding does not support negative indices               
  --> minicbor-tests/tests/encoding.rs:62:21                          
   |                                                                                                                                         
62 |     #[derive(Debug, Encode, Decode, PartialEq, Eq)]                                                                                     
   |                     ^^^^^^                                                                                                              
   |                                                                                                                                         
   = note: this error originates in the derive macro `Encode` (in Nightly builds, run with -Z macro-backtrace for more info)      
```

But when I switch over to `.orig` as in 0b7229bb34c6e65950de2e7f017b2cbe3df3cb43, I get this:

```
error: array encoding does not support negative indices                                                                                      
  --> minicbor-tests/tests/encoding.rs:67:9                                                                                                  
   |                               
67 |         #[n(-1)] c: Option<u8>                                                                                                          
   |         ^                                                                                                                               
           
```

I did not try triggering every single error that gets created, but generally, `.index.span()` can't really do better. (I don't even know how it gets the line for the derive right; my guess is that once an error without a span is raised from a derive, the derive's line gets set).

As the `.orig` does raise questions, the more self-evident version would be to use `syn::Error::new(field.span(), …)` -- done in the last commit by implementing `ToTokens` on the whole `Field` (which then delegates that to `.orig`). An alternative would be to extend the `Idx` enum to also carry around the spans, but I think that's overdoing it: the user is already pointed to the right annotation.

---

I think this addresses all the comments (in many cases, just by applying your suggestions).

If you're happy with the content I'd squash the changes to have more meaningful single commits per change rather than the back-and-forth it is now. (This entails rewriting the commit messages, which is why I didn't particularly signed off the fixup commits -- their commit messages get lost at squashing time anyway).

twittner commented:
> If you're happy with the content I'd squash the changes to have more meaningful single commits per change rather than the back-and-forth it is now [...]

Sounds good.

twittner commented on minicbor-derive/src/attrs/idx.rs:
```diff
@@ -24,12 +24,24 @@ impl Idx {
+    ///
+    /// This is used for pre-sorting items for serialization.
+    pub fn val_for_sorting(self) -> impl Ord {
+        (self.val() < 0, self.val().unsigned_abs())
```
I do not think this is actually producing a bytewise lexicographic order as it would sort `[-1, -10, -100]` as `[-1, -10, -100]`, however the correct order would be `[-100, -10, -1]` (`[9c, f6, ff]`). The correct approach would be to sort by `i32::to_be_bytes`. I suggest you remove this method and in `Fields` you do `fields.sort_unstable_by_key(|f| f.index.val().to_be_bytes());`.

chrysn commented:
Thanks for the thorough review.

I'm not sure I understand the "merging is blocked" reasoning which GitHub shows -- is this because merges have to be fast-forward? I.e., should I rebase this on top of the `develop` branch?

twittner commented:
Thanks, merged manually in 86958ee8de8e8d64cbf2a549f137a32742aaa64f.

twittner commented:
I changed to `i64` instead of `i32` for backwards compatibility.

chrysn commented:
Right, good catch -- I had discarded that because one doesn't simply do 2^31 long lists, but forgot to consider map keys.
